### PR TITLE
[PyTorch] Replace RecordFunction shouldRun callback with atomic bools

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56504 [PyTorch] Replace RecordFunction shouldRun callback with atomic bools**

Having callbacks registered but disabled via their
`shouldRun` callback defeats the `shouldRunRecordFunction`
optimization (no relation between the two things, despite the
shared prefix on the names) that aims to skip `RecordFunction`
construction.

This diff attempts to safely rectify this issue: we drop support for
`shouldRun` callbacks (this is bc-breaking; does anything use these
externally? do I need to add the support back and just stop using it
internally?), add support for enabling and disabling callbacks, and
(for global callbacks) make doing so thread-safe.

There is an interesting subtlety with `std::atomic` that came up: it
is neither copyable nor movable, which precludes putting it into
`std::vector`. I manually overrode this because the thread safety
reasons it is neither copyable nor movable don't apply here; we
already state that adding or removing callbacks (the operations that
might copy/move an atomic) are not thread-safe and should be done at
initialization time.

Differential Revision: [D27834944](https://our.internmc.facebook.com/intern/diff/D27834944/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27834944/)!

### CODE REVIEW

**Comment by swolchok** [comment]
ping

File: aten/src/ATen/record_function.cpp [hunk 1]
```diff
@@ -88,8 +89,64 @@ void set_record_function_tls_(const RecordFunctionTLS& tls) {
...
+static bool findAndRemoveCallback(
+    RecordFunctionCallbacks& cbs, CallbackHandle handle) {
+  auto it = std::find_if(
+      cbs.begin(), cbs.end(),
+      [handle](
+          const auto& el) {
+        return el.handle == handle;
+      });
+  if (it != cbs.end()) {
+    // We do not need to try to call releaseRecordAllFunctionsHere
```
Comment (ilia-cher) [inline aten/src/ATen/record_function.cpp:135]: thanks for adding a clarification, super nit: space

File: aten/src/ATen/record_function.cpp [hunk 2]
```diff
@@ -88,8 +89,64 @@ void set_record_function_tls_(const RecordFunctionTLS& tls) {
...
+enum class ToggledCallbackResult {
+  NotFound,
+  FoundButNotToggled,
+  FoundAndToggled,
+};
+
+template <typename RecordFunctionCallbacks>
+static ToggledCallbackResult findAndToggleCallback(
+    RecordFunctionCallbacks& cbs, CallbackHandle handle, bool enabled) {
+  auto it = std::find_if(
```
Comment (ilia-cher) [inline aten/src/ATen/record_function.cpp:101]: yeah, we could use the fact that the vector is sorted, but this indeed is not supposed to be called in the hot path, also we don't have that many callbacks anyways

File: aten/src/ATen/record_function.cpp [hunk 3]
```diff
@@ -525,7 +606,7 @@ void bumpRecordAllFunctions() {
 }
 
 void releaseRecordAllFunctions() {
-  TORCH_CHECK(global_record_all_functions_.fetch_sub(1, std::memory_order_relaxed) >= 0);
+  TORCH_CHECK(global_record_all_functions_.fetch_sub(1, std::memory_order_relaxed) > 0);
```
Comment (ilia-cher) [inline aten/src/ATen/record_function.cpp:609]: why do we disallow zero here?

**Review by ilia-cher — COMMENTED** [review]
Thanks! Let's remove shouldRun in favor of a simpler boolean flag

File: aten/src/ATen/record_function.cpp (hunk 3 above)
Comment (swolchok) [reply aten/src/ATen/record_function.cpp:609]: because fetch_sub returns the old value. if the old value was 0, the new value is -1, which is incorrect.

**Review by ilia-cher — APPROVED** [review]

**Pull request closed without merging**